drivers/flash/flash_mspi_nor.c

```diff
@@ -0,0 +1,587 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT jedec_mspi_nor
+
+#include <zephyr/drivers/flash.h>
+#include <zephyr/drivers/mspi.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/pm/device_runtime.h>
+
+#include "spi_nor.h"
+#include "jesd216.h"
+#include "flash_priv.h"
+
+LOG_MODULE_REGISTER(flash_mspi_nor, CONFIG_FLASH_LOG_LEVEL);
+
+struct flash_mspi_nor_data {
+	struct k_sem acquired;
+	struct mspi_xfer_packet packet;
+	struct mspi_xfer xfer;
+};
+
+struct flash_mspi_nor_config {
+	const struct device *bus;
+	struct mspi_dev_id mspi_id;
+	struct mspi_dev_cfg mspi_cfg;
+	struct mspi_xip_cfg xip_cfg;
+	struct gpio_dt_spec reset;
+	uint32_t flash_size;
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+	struct flash_pages_layout layout;
+#endif
+	uint8_t jedec_id[SPI_NOR_MAX_ID_LEN];
+};
+
+static int acquire(const struct device *dev)
+{
+	const struct flash_mspi_nor_config *dev_config = dev->config;
+	struct flash_mspi_nor_data *dev_data = dev->data;
+	int rc;
+
+	k_sem_take(&dev_data->acquired, K_FOREVER);
+
+	rc = pm_device_runtime_get(dev_config->bus);
+	if (rc < 0) {
+		LOG_ERR("pm_device_runtime_get() failed: %d", rc);
+		return rc;
+	}
+
+	rc = mspi_dev_config(dev_config->bus, &dev_config->mspi_id,
+			     MSPI_DEVICE_CONFIG_ALL, &dev_config->mspi_cfg);
+	if (rc < 0) {
+		LOG_ERR("mspi_dev_config() failed: %d", rc);
+		return rc;
+	}
+
+	return 0;
+}
+
+static void release(const struct device *dev)
+{
+	const struct flash_mspi_nor_config *dev_config = dev->config;
+	struct flash_mspi_nor_data *dev_data = dev->data;
+
+	(void)pm_device_runtime_put(dev_config->bus);
+
```

Comment:
Here are the reasons when I considered it
1. There are no precedence in all the driver API in zephyr to have explicit acquire and release. 
2. I think it is too cumbersome to have those explicitly. Putting the functionality into mspi_dev_config & mspi_get_channel_status is really convenient and should work seamlessly (invisible to the device driver but should be understood clearly by controller developers).
3. `mspi_get_channel_status` should not simply release controller access upon called. My apologies, I should not have said that. It qeuries whether there are still ongoing transfers on the controller. And when there were no more transfers, it releases controller access naturely. There is no need for a k_timeout_t because it is in each transfer already. As for controller error status, they should be handled prior return. I don't know if there are cases where the error can't be handled prior return.
4. The real purpose of this is to be able to use software multi-target in device drivers. Now you can choose not to support it in controller driver, but we need it in device drivers for maximum compatibility.

I had requested to conslidate the API return values on my [first PR ](https://github.com/zephyrproject-rtos/zephyr/pull/71769) for at least two weeks and yet there were no disagreement or feedback.... Thus the return value for `mspi_get_channel_status` is either 0 for success or -EBUSY for when the controller is still busy processing transfer.

As I wrote in the doc: the mspi_dev_config implementation should do the following. Naturally, you would have to handle the release of the controller access when configuration/reconfiguration failed.

> Call [mspi_dev_config()](https://docs.zephyrproject.org/latest/doxygen/html/group__mspi__configure__api.html#gaa3c1eae8b3000c9bafcc26f14a8beb8b) with device specific hardware settings obtained from device datasheets.
> 
> The [mspi_dev_cfg](https://docs.zephyrproject.org/latest/doxygen/html/structmspi__dev__cfg.html) should be filled by device tree and helper macro [MSPI_DEVICE_CONFIG_DT](https://docs.zephyrproject.org/latest/doxygen/html/group__mspi__devicetree.html#gaa8e730d85dede3d1e5dc9f69f30098b2) can be used.
> 
> The controller driver should then validate the members of [mspi_dev_cfg](https://docs.zephyrproject.org/latest/doxygen/html/structmspi__dev__cfg.html) to prevent incorrect usage.
> 
> The controller driver should implement a mutex to protect from accidental access.
> 
> The controller driver may also switch between different devices based on [mspi_dev_id](https://docs.zephyrproject.org/latest/doxygen/html/structmspi__dev__id.html).

Admittedly, the while loop in the atxp032 driver is not optimal yet I'm not sure how much trouble it is going to be considering the flash API either requires blocking or synchronous only. The reason why it is a while loop is becuase the acquire and release function in the device drivers are typically have return type void and also for the fact that return values for APIs were not discussed. So I'm not sure how to handle the errors in the flash API context. 

If you or anyone want to discuss the return values for the APIs, I can open up an issue.

Reply:
I think the only time we would need a  k_timeout_t is when software multi-target is enabled and when a different device tries to acquire controller when the current one is still transferring. But this is already taken care of with the mutex in mspi_dev_config. For releasing, you don't need it unless there has been async operation which the flash API don't appreciate.